lms/djangoapps/course_goals/models.py

```diff
@@ -40,15 +40,13 @@ class Meta:
         unique_together = ("user", "course_key")
 
 
-@receiver(models.signals.post_save, sender=CourseEnrollment, dispatch_uid="create_course_goal_on_enroll")
+@receiver(models.signals.post_save, sender=CourseEnrollment, dispatch_uid="update_course_goal_on_enroll_change")
 def set_course_goal_verified(sender, instance, **kwargs):  # pylint: disable=unused-argument, invalid-name
-    """Set the course goal to certify when the user enrolls as a verified user. """
-    ## If the user enrolls in a verified state or upgrades, create a course goal
+    """
+    Set the course goal to certify when the user enrolls as a verified user.
+    Remove the course goal when the user's enrollment is no longer active.
+    """
     if instance.mode == 'verified':
         add_course_goal(instance.user, instance.course_id, 'certify')
-
-
-@receiver(models.signals.post_delete, sender=CourseEnrollment, dispatch_uid="remove_course_goal_on_unenroll")
-def remove_course_goal_on_unenroll(sender, instance, **kwargs):  # pylint: disable=unused-argument, invalid-name
-    """Remove course goal on unenroll. """
-    remove_course_goal(instance.user, instance.course_id)
+    if not instance.is_active:
```

Comment:
If you want to combine these, I'd put this one first (and possible use if/else), so if it happens to be a verified user that is unenrolling you don't add and then remove.

Reply:
Switched